Correct errors in the reference of extern functions definitions and declarations (#652)
rust-lang-nursery/reference: src/items/functions.md

sufficient context to determine the type parameters. For example,
`mem::size_of::<u32>() == 4`.

## Extern function qualifier

The `extern` function qualifier allows providing function _definitions_ that can

> be called with a particular ABI:

Reviewer: Hmm, this seems to imply it's specific to definitions but the exact same syntax with the same default also applies to `fn` pointers.

Author: Not the exact same (e.g. function pointers have no identifier or body), but I see your point. Stretching this out a bit more, an also similar syntax applies to declarations in extern blocks. 

I think that the "extern block item" and the "function item" sections should just tangentially mention `extern "ABI"`, and that a different section describing the "extern qualifier" should be used to explain the whole thing. As it is done right now, we have links back and forth between both sections for no good reason.

Reviewer 2: IMO, punt this change to another PR. This one is getting a lot of back and forth, and I think we should setting with an improved base to work on.

That said, for that section, something like this could work:

```md
The `extern` qualifier specifies the [<abbr title="Application Binary Interface>ABI</abbr>][ABI] for calling the functions they apply to. It is applicable to ...

[ABI]: https://en.wikipedia.org/ABI
```

Author: So I'm going to leave this as is for now . Even if it still isn't 100% clear, it is still an improvement over what we have. I've opened #655 to track my suggestion for improving this further by just explaining the `extern` qualifier for all users somewhere else, and just referencing it from the function item and extern block item sections. 

If you have any better ideas about how to explain this more clearly, we can discuss those in that issue.

```rust,ignore
extern "ABI" fn foo() { ... }
```

These are often used in combination with [external block] items which provide
function _declarations_ that can be used to call functions without providing
their _definition_:

```rust,ignore
extern "ABI" {
  fn foo(); /* no body */
}
unsafe { foo() }
```

When `"extern" Abi?*` is omitted from `FunctionQualifiers` in function items,
the ABI `"Rust"` is assigned. For example:

```rust
fn foo() {}
```

is equivalent to:

```rust
extern "Rust" fn foo() {}
```

> Functions in Rust can be called by foreign code, and using an ABI that

Reviewer 2: This is a run-on sentence, or at least feels like one. Also, "Rust's functions". And what does "foreign code" here mean differently than code from other programming languages?

Reviewer 3: Rust code exposed with a C ABI.

Author: I'm not sure how to improve this.

}
unsafe { foo() }
```

When `"extern" Abi?*` is omitted from `FunctionQualifiers` in function items,
the ABI `"Rust"` is assigned. For example:

```rust
fn foo() {}
```

is equivalent to:

```rust
extern "Rust" fn foo() {}
```

Functions in Rust can be called by foreign code, and using an ABI that
differs from Rust allows, for example, to provide functions that can be 
called from other programming languages like C:

```rust
// Declares a function with the "C" ABI
extern "C" fn new_i32() -> i32 { 0 }

// Declares a function with the "stdcall" ABI
# #[cfg(target_arch = "x86_64")]
extern "stdcall" fn new_i32_stdcall() -> i32 { 0 }
```

Just as with [external block], when the `extern` keyword is used and the `"ABI` 
is omitted, the ABI used defaults to `"C"`. That is, this:

```rust
extern fn new_i32() -> i32 { 0 }
let fptr: extern fn() -> i32 = new_i32;
```

is equivalent to:

```rust
extern "C" fn new_i32() -> i32 { 0 }
let fptr: extern "C" fn() -> i32 = new_i32;
```

Functions with an ABI that differs from `"Rust"` do not support unwinding in the
exact same way that Rust does. Therefore, unwinding past the end of functions
with such ABIs causes the process to abort.

> **Non-normative note**: The LLVM backend of the current Rust implementation

Reviewer 3: ```suggestion
> **Note**: The LLVM backend of the `rustc` implementation
```

## Extern functions

Extern function _definitions_ allow defining functions that can be called 
with a particular ABI:

> ```rust,norun

Reviewer 4: `norun` is not a valid attribute, it is `no_run`. But it should not be necessary if it is made into valid syntax.

> extern "ABI" fn foo() { ... }

Reviewer 4: There's a CI error here.  I would just show a simple example like:

```suggestion
extern "C" fn foo() { /*...*/ }
```

Author: @ehuss I think using `"ABI"` here is better for exposition purposes. I tried to avoid running this with `norun`, but that did not work out. I'll fix that.

Author: This should be fixed in the latest commit, so I'm resolving this issue. 

Please feel free to re-open if you disagree with using `"ABI"` for exposition purposes.

```

An extern function _declaration_ via an [external block] can be used to
provide an item for these functions that can be called by Rust code without
providing their definition. 

The default ABI of Rust functions like `fn foo() {}` is `"Rust"`. While we
abbreviate the type of Rust functions like `foo` as `fn()`, this is actually a 
synonym for `extern "Rust" fn()`. That is, this:

```rust
fn foo() {}
```

> is identical to 

Reviewer 3: ```suggestion
is equivalent to:
```

```rust
extern "Rust" fn foo() {}
```

Functions in Rust can be called by foreign code, and using an ABI that
differs from Rust allows, for example, to provide functions that can be 
called from other programming languages like C:

```rust
// Declares a function with the "C" ABI
extern "C" fn new_i32() -> i32 { 0 }

// Declares a function with the "stdcall" ABI
# #[cfg(target_arch = "x86_64")]
extern "stdcall" fn new_i32_stdcall() -> i32 { 0 }
```

Just as with [external block], when the `extern` keyword is used and the `"ABI` 

> is omitted, the ABI used defaults to `"C"`. That is, this

Reviewer 3: ```suggestion
is omitted, the ABI used defaults to `"C"`. That is, this:
```

```rust
extern fn new_i32() -> i32 { 0 }
let fptr: extern fn() -> i32 = new_i32;
```

> is identical to

Reviewer 3: ```suggestion
is equivalent to:
```

```rust
extern "C" fn new_i32() -> i32 { 0 }
let fptr: extern "C" fn() -> i32 = new_i32;
```

> Since functions with an ABI that differs from `"Rust"` do not support 

Reviewer 3: ```suggestion
Functions with an ABI that differs from `"Rust"` do not support 
```

> unwinding in the exact same way that Rust does, unwinding past the end 

Reviewer 3: ```suggestion
unwinding in the exact same way that Rust does. Therefore, unwinding past the end 
```

> of functions with such ABIs causes the process to abort. In LLVM, this is

Reviewer 3: What LLVM does is not normative and should at most be left as a note.

Author: Moved this to a non-normative note.

implemented by executing an illegal instruction.

> Some ABIs that are identical to `"Rust"` are:

Reviewer 3: Remove this bit; these are all unstable.

Author: @Centril where would be the right place to document that these are all equivalent to the `"Rust"` ABI ? The "unstable" Reference? Maybe I can add that there. 

I don't know if it would be worth it to hyperlink that from here, or to just have a small mention here that there are some unstable ABIs that are also equivalent to `"Rust"`. 

Thoughts?

Author: @Centril I've added a "Non-normative note" that mentions that there are some unstable ABIs that are equivalent to `"Rust"`, and that refers readers to the ABI section of the reference - which for some reason mentions these ABIs as well without saying that these are unstable - and the "Unstable Book".

Reviewer 2: We want to make zero reference to unstable parts of the language. Not even notes.

Those ABIs should be documented in the rustc book.

Author: Shall I remove mentions to these both from this section and the extern block section then?

Reviewer 3: Seems like a good idea yeah.

```

Just as with [external block], when the `extern` keyword is used and the `"ABI` 
is omitted, the ABI used defaults to `"C"`. That is, this:

```rust
extern fn new_i32() -> i32 { 0 }
let fptr: extern fn() -> i32 = new_i32;
```

is equivalent to:

```rust
extern "C" fn new_i32() -> i32 { 0 }
let fptr: extern "C" fn() -> i32 = new_i32;
```

Functions with an ABI that differs from `"Rust"` do not support unwinding in the
exact same way that Rust does. Therefore, unwinding past the end of functions
with such ABIs causes the process to abort.

**Non-normative note**: The LLVM backend of the current Rust implementation
aborts the process by executing an illegal instruction.

**Non-normative note**: There are other ABIs available in unstable Rust that are

> equivalent to the `"Rust"` ABI, e.g.,

Reviewer 5: Are they actually equivalent to `extern "Rust"`, though? I know for example that pointers to `extern "rust-intrinsic"` functions cannot be formed, since they have nowhere to point to.

Author: I see your point. The pointers can be formed (e.g. Rust programs that do this type check), but the compiler ICEs (this detects bugs in libcore, etc.). 

Maybe you are interested in [this comment](https://github.com/rust-lang-nursery/reference/pull/652#discussion_r314206468).

Reviewer 5: I always assumed that because intrinsics are always inlined into the caller, they don't have their own distinct addresses and it's meaningless to speak of pointers to them. The compiler ICEs because it cannot generate code which forms such pointers, even when all types check out; in other words, it's not a deliberate feature (see also rust-lang/rust#15694). Am I wrong here?

Author: I'm not sure. I don't know what happens when you create a function pointer to, e.g., `llvm.assume` (does one get an LLVM error?), but for something like `core::intrinsic::sqrtf32` which emits `llvm.sqrt` creating a function pointer to it should work just fine. I'm not sure, but it might depend on the intrinsic.

[`export_name`]: ../abi.md#the-export_name-attribute
[`link_section`]: ../abi.md#the-link_section-attribute
[`no_mangle`]: ../abi.md#the-no_mangle-attribute
[external_block_abi]: external-blocks.md#abi
[Unstable Book]: https://doc.rust-lang.org/unstable-book

> [rust_intrinsic]: https://doc.rust-lang.org/unstable-book/language-features/intrinsics.html?highlight=rust-intrin#intrinsics

Reviewer 3: Links need some garbage collection now.
